### DISCUSSION TITLE
Getting started connecting to sqlite db

### DISCUSSION THREAD

**Discussion by 27concepts** [kkharji/sqlite.lua#148] [Q&A]
Hi, I am lost as to how to get started here.

I have a sqlite db and I just want to query it. I am able to connect to it via DB Browser for SQLite, but no idea how to go about it using lua code.

I moved the sqlite folder in the same folder as my lua code and added local has_sqlite, sqlite = pcall(require, "sqlite") that I found in the other discussion. It doesn't work. 

Help?

Thanks.

**Comment by kkharji** [comment]
Hey @27concepts, sorry to hear that.

If you installed sqlite.lua then it the access part should be dealt with. Now, if you have a database filled with data, I suggest you first back it up, just in case anything goes wrong, or possibly of being locked due to write ops (see [using sqlite with firefox database](https://github.com/dawsers/telescope-firefox.nvim/blob/main/lua/telescope/_extensions/firefox/firefox.lua))

There are two ways in which you can start communicating with database and all of theme are simple and straightforward :

### 1. quick and dirty
Open the db, execute statements and close it
```lua
local items = sqlite:with_open("/path/to/db", function(mydb) 
     -- see `: h sqlite_query_select` && `:h sqlite.db:select`
    return mydb:select("tablename", { where = {  key = "value" }})
    -- just evaluate a sql statement
   return db:eval("select * from tablename where x = ?", 1) -- 
end)
```

### 2. structured 

This one is much more elegant, stable and super lazy, with auto closing

```lua
local mydb = sqlite {
   uri = "path/to/db",
   tablename = {
   -- ... keys and their types
  }
}

local tablename = mydb.tablename

tablename:count() -- number of elements
```

Now you can see `:h sqlite.tbl.lua`  to see all the methods you can do with table object.

References:
1. [Help docs](https://github.com/kkharji/sqlite.lua/blob/master/doc/sqlite.txt) `:h sqlite.readme`, the api is well documented and everything possible can be found there
2. [Tests](https://github.com/kkharji/sqlite.lua/tree/master/test/auto), sometimes the best way to know how something behave is to look at the tests. ..

Does this help?

**Reply by 27concepts** [reply]
Thank you for replying! 

Thing is, I am not sure I installed sqlite.lua correctly. What exactly needs to be done to "install" it? I downloaded the sqlite.lua-master.zip file from Github here and unzipped it in the folder of where my .lua file is.

Anything else needs to be done? How can I confirm it's installed correctly?

**Reply by kkharji** [reply]
Oh, hmmm sqlite requires [luv](https://luarocks.org/modules/creationix/luv) and LuaJIT `lua -v`.

If you have luv and running on luajit then copying it should just work. 

btw what kind of error you get from doing require'sqlite' and the code snippet above?

**Reply by 27concepts** [reply]
Ok, then I didn't install anything regarding LuaJIT. I am looking for a solution that is "portable", that is, dependent solely on a .dll or something that doesn't require installation on another system.

**Reply by kkharji** [reply]
I suppose luv dependency can be removed, see how it is used https://github.com/kkharji/sqlite.lua/search?q=luv. If you have some time to replace them with built in method, I can merge it

- https://github.com/kkharji/sqlite.lua/blob/b3de8f58181e8b1a2067bd28335037c559e15dd2/lua/sqlite/utils.lua#L106
- https://github.com/kkharji/sqlite.lua/blob/56c5aacd5e31496d9b3cd3d1b0e570bb9a65d35b/lua/sqlite/defs.lua#L18
- https://github.com/kkharji/sqlite.lua/blob/ca8233f8cb09b9adc7ea11f81ab903154ce07e86/lua/sqlite/helpers.lua#L76
- https://github.com/kkharji/sqlite.lua/blob/78ff09f026f4b36971b8ee6cd45ffd7a3865d49e/lua/sqlite/tbl/cache.lua#L65
